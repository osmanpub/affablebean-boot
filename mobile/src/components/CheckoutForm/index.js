import React, { Component } from "react";
import PropTypes from "prop-types";
import { purchaseOrder } from "../../net/checkout";
import { Button, StyleSheet, Text, TextInput, View } from "react-native";
import { validateField } from "../../utils";

export default class CheckoutForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      address: "",
      creditcard: "",
      email: "",
      name1: "",
      phone: "",
      validForm: true
    };
  }

  onChange = (name, value) => {
    this.setState({
      [name]: value
    });
  };

  purchaseItems = () => {
    const { dispatch, cart } = this.props;
    let validForm = true;

    if (!validateField(this.state.address.length, 8, 45)) {
      validForm = false;
    }

    if (!validateField(this.state.creditcard.length, 16, 19)) {
      validForm = false;
    }

    if (!validateField(this.state.email.length, 8, 45)) {
      validForm = false;
    }

    if (!validateField(this.state.name1.length, 8, 45)) {
      validForm = false;
    }

    if (!validateField(this.state.phone.length, 8, 30)) {
      validForm = false;
    }

    if (validForm) {
      dispatch(
        purchaseOrder({
          cart: cart,
          form: this.state
        })
      );
    } else {
      this.setState({ validForm: false });
    }
  };

  render() {
    const { cart } = this.props;
    const surcharge = 3;

    const styles = StyleSheet.create({
      error: {
        backgroundColor: "red",
        color: "white",
        fontWeight: "bold",
        padding: 8
      },
      textInput: {
        borderBottomColor: "black",
        borderBottomWidth: 1,
        padding: 8
      }
    });

    const validForm = this.state.validForm;

    return (
      <View>
        <Text style={{ padding: 8 }} />
        <View style={{ backgroundColor: "#f7f7e9", padding: 16 }}>
          <Text>
            In order to purchase the items in your shopping cart, please provide
            us with the following information:
          </Text>
          <TextInput
            onChangeText={name => this.onChange("name1", name)}
            placeholder="Enter your name"
            style={styles.textInput}
            value={this.state.name1}
          />
          {!validForm && (
            <Text style={styles.error}>
              Name should be between 8 and 45 characters
            </Text>
          )}
          <TextInput
            onChangeText={email => this.onChange("email", email)}
            placeholder="Enter your email address"
            style={styles.textInput}
            value={this.state.email}
          />
          {!validForm && (
            <Text style={styles.error}>
              Email should be between 8 and 45 characters
            </Text>
          )}
          <TextInput
            onChangeText={phone => this.onChange("phone", phone)}
            placeholder="Enter your phone number"
            style={styles.textInput}
            value={this.state.phone}
          />
          {!validForm && (
            <Text style={styles.error}>
              Phone number should be between 8 and 30 characters
            </Text>
          )}
          <TextInput
            onChangeText={address => this.onChange("address", address)}
            placeholder="Enter your home address"
            style={styles.textInput}
            value={this.state.address}
          />
          {!validForm && (
            <Text style={styles.error}>
              Address should be between 8 and 45 characters
            </Text>
          )}
          <TextInput
            onChangeText={creditcard => this.onChange("creditcard", creditcard)}
            placeholder="Enter your credit card number"
            style={styles.textInput}
            value={this.state.creditcard}
          />
          {!validForm && (
            <Text style={styles.error}>
              Credit card number should be between 16 and 19 characters
            </Text>
          )}
        </View>
        <View style={{ alignItems: "center" }}>
          <Text style={{ paddingTop: 16 }}>
            Subtotal: &euro; {cart.subtotal.toFixed(2)}
          </Text>
          <Text>Surcharge: &euro; {surcharge.toFixed(2)}</Text>
          <Text style={{ fontWeight: "bold" }}>
            Total: &euro; {(cart.subtotal + surcharge).toFixed(2)}{" "}
          </Text>
          <Text style={{ marginTop: 8 }} />
          <Button onPress={this.purchaseItems} title="Purchase" />
          {!validForm && (
            <Text style={styles.error}>Check your form for errors!</Text>
          )}
        </View>
      </View>
    );
  }
}

CheckoutForm.propTypes = {
  cart: PropTypes.object.isRequired
};
