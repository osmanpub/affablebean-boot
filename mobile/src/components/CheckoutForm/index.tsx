import React from 'react';
import {Controller, useForm} from 'react-hook-form';
import {Button, StyleSheet, Text, TextInput, View} from 'react-native';
import {connect, useDispatch} from 'react-redux';
import {Cart} from '../../interfaces/cart';
import {FormErrors} from '../../interfaces/ui';
import {purchaseOrder} from '../../net/checkout';
import {RootState} from '../../redux';
import {setFormErrors} from '../../redux/ui';

type Props = {
  cart: Cart;
  formErrors: Array<FormErrors>;
  setFormErrors: Function;
};

type FormData = {
  address: string;
  creditCard: string;
  email: string;
  name: string;
  phone: string;
};

function CheckoutForm(props: Props) {
  const {cart, formErrors, setFormErrors} = props;
  const dispatch = useDispatch();
  const {control, handleSubmit, errors} = useForm<FormData>();

  const onChange = (args: Array<any>) => {
    return {
      value: args[0].nativeEvent.text,
    };
  };

  const onSubmit = handleSubmit(({address, creditCard, email, name, phone}) => {
    dispatch(purchaseOrder({address, creditCard, email, name, phone}));
  });

  if (formErrors && Array.isArray(formErrors) && formErrors.length > 0) {
    let msg = '';

    formErrors.forEach(error => {
      msg += `Field "${error.param}" with value "${error.value}" has the following problem:\n"${error.msg}"`;
    });

    setFormErrors([]);

    alert(
      `There was a problem processing your order.\nPlease correct the following errors:\n${msg}`,
    );
  }

  return (
    <View>
      {/* <Text style={{padding: 8}} /> */}
      <View style={{backgroundColor: '#f7f7e9', padding: 16}}>
        <Text>
          In order to purchase the items in your shopping cart, please provide
          us with the following information:
        </Text>
        <Controller
          as={<TextInput />}
          control={control}
          name="name"
          onChange={onChange}
          rules={{required: true, minLength: 3, maxLength: 64}}
          placeholder="Enter your name"
          style={styles.textInput}
        />
        {errors.name && (
          <Text style={styles.error}>
            Name should be between 3 and 64 characters
          </Text>
        )}
        <Controller
          as={<TextInput />}
          control={control}
          name="email"
          onChange={onChange}
          rules={{required: true, minLength: 8, maxLength: 32}}
          placeholder="Enter your email address"
          style={styles.textInput}
        />
        {errors.email && (
          <Text style={styles.error}>
            Email should be between 8 and 32 characters
          </Text>
        )}
        <Controller
          as={<TextInput />}
          control={control}
          name="phone"
          onChange={onChange}
          rules={{required: true, minLength: 8, maxLength: 32}}
          placeholder="Enter your phone number"
          style={styles.textInput}
        />
        {errors.phone && (
          <Text style={styles.error}>
            Phone should be between 8 and 32 characters
          </Text>
        )}
        <Controller
          as={<TextInput />}
          control={control}
          name="address"
          onChange={onChange}
          rules={{required: true, minLength: 8, maxLength: 256}}
          placeholder="Enter your address"
          style={styles.textInput}
        />
        {errors.address && (
          <Text style={styles.error}>
            Address should be between 8 and 256 characters
          </Text>
        )}
        <Controller
          as={<TextInput />}
          control={control}
          name="creditCard"
          onChange={onChange}
          rules={{required: true, minLength: 16, maxLength: 19}}
          placeholder="Enter your credit card number"
          style={styles.textInput}
        />
        {errors.creditCard && (
          <Text style={styles.error}>
            Credit card number should be between 16 and 19 characters
          </Text>
        )}
      </View>
      <View style={{alignItems: 'center'}}>
        <Text style={{paddingTop: 16}}>
          Subtotal: &euro; {cart.subtotal.toFixed(2)}
        </Text>
        <Text>Surcharge: &euro; {surcharge.toFixed(2)}</Text>
        <Text style={{fontWeight: 'bold'}}>
          Total: &euro; {(cart.subtotal + surcharge).toFixed(2)}{' '}
        </Text>
        <Text style={{marginTop: 8}} />
        <Button onPress={handleSubmit(onSubmit)} title="Purchase" />
      </View>
    </View>
  );
}

const surcharge = 3;

const styles = StyleSheet.create({
  error: {
    color: 'red',
    fontWeight: 'bold',
  },
  textInput: {
    borderBottomColor: 'black',
    borderBottomWidth: 1,
    padding: 8,
  },
});

const mapStateToProps = (state: RootState) => ({
  cart: state.cart,
  formErrors: state.ui.formErrors,
});

const mapDispatchToProps = {
  setFormErrors,
};

export default connect(mapStateToProps, mapDispatchToProps)(CheckoutForm);
