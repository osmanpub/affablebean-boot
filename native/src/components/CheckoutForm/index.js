import React, { Component } from "react";
import PropTypes from "prop-types";
import { purchaseOrder } from "../../net/checkout";
import { Button, StyleSheet, Text, TextInput, View } from "react-native";
import { validateField } from "../../utils";

export default class CheckoutForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      address: "",
      creditcard: "",
      email: "",
      name1: "",
      phone: ""
    };
  }

  onChange = (name, value) => {
    this.setState({
      [name]: value
    });
  };

  purchaseItems = () => {
    const { dispatch, cart } = this.props;
    let validForm = true;

    if (!validateField(this.addressInputRef, this.addressErrorRef, 8, 45)) {
      validForm = false;
    }

    if (!validateField(this.ccInputRef, this.ccErrorRef, 16, 19)) {
      validForm = false;
    }

    if (!validateField(this.emailInputRef, this.emailErrorRef, 8, 45)) {
      validForm = false;
    }

    if (!validateField(this.nameInputRef, this.nameErrorRef, 8, 45)) {
      validForm = false;
    }

    if (!validateField(this.phoneInputRef, this.phoneErrorRef, 8, 30)) {
      validForm = false;
    }

    if (validForm) {
      dispatch(
        purchaseOrder({
          cart: cart,
          form: this.state
        })
      );
    }
  };

  render() {
    const { cart } = this.props;
    const surcharge = 3;

    const styles = StyleSheet.create({
      textInput: {
        borderBottomColor: "black",
        borderBottomWidth: 1,
        padding: 8
      }
    });

    return (
      <View>
        <Text style={{ padding: 8 }} />
        <View style={{ backgroundColor: "#f7f7e9", padding: 16 }}>
          <Text>
            In order to purchase the items in your shopping cart, please provide
            us with the following information:
          </Text>
          <TextInput
            onChangeText={name => this.onChange("name1", name)}
            placeholder="Enter your name"
            style={styles.textInput}
            value={this.state.name1}
          />
          <TextInput
            onChangeText={email => this.onChange("email", email)}
            placeholder="Enter your email address"
            style={styles.textInput}
            value={this.state.email}
          />
          <TextInput
            onChangeText={phone => this.onChange("phone", phone)}
            placeholder="Enter your phone number"
            style={styles.textInput}
            value={this.state.phone}
          />
          <TextInput
            onChangeText={address => this.onChange("address", address)}
            placeholder="Enter your home address"
            style={styles.textInput}
            value={this.state.address}
          />
          <TextInput
            onChangeText={creditcard => this.onChange("creditcard", creditcard)}
            placeholder="Enter your credit card number"
            style={styles.textInput}
            value={this.state.creditcard}
          />
        </View>
        <View style={{ alignItems: "center" }}>
          <Text style={{ paddingTop: 16 }}>
            Subtotal: &euro; {cart.subtotal.toFixed(2)}
          </Text>
          <Text>Surcharge: &euro; {surcharge.toFixed(2)}</Text>
          <Text style={{ fontWeight: "bold" }}>
            Total: &euro; {(cart.subtotal + surcharge).toFixed(2)}{" "}
          </Text>
          <Text style={{ marginTop: 8 }} />
          <Button onPress={this.purchaseItems} title="Purchase" />
        </View>
      </View>
    );
  }
}

CheckoutForm.propTypes = {
  cart: PropTypes.object.isRequired
};
