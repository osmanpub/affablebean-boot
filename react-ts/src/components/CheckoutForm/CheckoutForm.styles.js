import styled from "styled-components";

export const InfoBox = styled.div`
  margin-left: 150px;
  /*padding: 10px 25px 0 460px*/
`;

export const PriceBox = styled.table`
  padding: 10px;
  border: 1px solid #c3e3e0;
  background-color: #c3e3e0;
  border-radius: 4px;
`;

export const PriceBoxTd = styled.td`
  text-align: right;
  padding: 5px;
`;

export const PriceBoxSubTotalTd = styled.td`
  width: 80px;
  font-weight: bold;
  text-align: right;
  padding: 5px;
`;
