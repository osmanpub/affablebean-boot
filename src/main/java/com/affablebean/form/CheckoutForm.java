package com.affablebean.form;

import javax.validation.constraints.Email;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;

public class CheckoutForm {

	@Email
	@NotNull
	@Size(min = 8, max = 64)
	private String email;

	@NotNull
	@Size(min = 3, max = 64)
	private String name;

	@NotNull
	@Size(min = 8, max = 32)
	private String phone;

	@NotNull
	@Size(min = 8, max = 256)
	private String address;

	@NotNull
	@Size(min = 2, max = 2)
	private String cityRegion = "NY"; // hard coded for now :(

	@NotNull
	@Size(min = 16, max = 19)
	private String creditCard;

	public String getEmail() {
		return email;
	}

	public void setEmail(String email) {
		this.email = email;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getPhone() {
		return phone;
	}

	public void setPhone(String phone) {
		this.phone = phone;
	}

	public String getAddress() {
		return address;
	}

	public void setAddress(String address) {
		this.address = address;
	}

	public String getCityRegion() {
		return cityRegion;
	}

	public void setCityRegion(String cityRegion) {
		this.cityRegion = cityRegion;
	}

	public String getCreditCard() {
		return creditCard;
	}

	public void setCreditCard(String creditCard) {
		this.creditCard = creditCard;
	}

	@Override
	public String toString() {
		return "CheckoutForm [email=" + email + ", name=" + name + ", phone=" + phone + ", address=" + address
				+ ", cityRegion=" + cityRegion + ", creditcard=" + creditCard + "]";
	}

}
