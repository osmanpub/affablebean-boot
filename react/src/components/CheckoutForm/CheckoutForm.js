import React, { Component } from "react";
import { purchaseOrder } from "../../net/checkout";
import {
  InfoBox,
  PriceBox,
  PriceBoxTd,
  PriceBoxSubTotalTd
} from "./CheckoutForm.styles";
import "./CheckoutForm.css";
import { validateField } from "../../utils";

export class CheckoutForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      address: "",
      creditcard: "",
      email: "",
      name1: "",
      phone: ""
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);

    this.nameErrorRef = React.createRef();
    this.nameInputRef = React.createRef();

    this.emailErrorRef = React.createRef();
    this.emailInputRef = React.createRef();

    this.phoneErrorRef = React.createRef();
    this.phoneInputRef = React.createRef();

    this.addressErrorRef = React.createRef();
    this.addressInputRef = React.createRef();

    this.ccErrorRef = React.createRef();
    this.ccInputRef = React.createRef();
  }

  handleChange(event) {
    const input = event.target;
    this.setState({
      [input.name]: input.value
    });
  }

  handleSubmit(event) {
    event.preventDefault();
    const { dispatch, cart } = this.props;
    let validForm = true;

    if (!validateField(this.nameInputRef, this.nameErrorRef, 8, 45)) {
      validForm = false;
    }

    if (!validateField(this.emailInputRef, this.emailErrorRef, 8, 45)) {
      validForm = false;
    }

    if (!validateField(this.phoneInputRef, this.phoneErrorRef, 8, 30)) {
      validForm = false;
    }

    if (!validateField(this.addressInputRef, this.addressErrorRef, 8, 45)) {
      validForm = false;
    }

    if (!validateField(this.ccInputRef, this.ccErrorRef, 16, 19)) {
      validForm = false;
    }

    if (validForm) {
      dispatch(
        purchaseOrder({
          cart: cart,
          form: this.state
        })
      );
    }
  }

  render() {
    const { cart } = this.props;
    const surcharge = 3;

    return (
      <div>
        <div className="singleColumn">
          <h2>checkout</h2>
          <p>
            In order to purchase the items in your shopping cart, please provide
            us with the following information:
          </p>
          <br />
          <form className="form-horizontal" onSubmit={this.handleSubmit}>
            <div className="form-group">
              <label htmlFor="name" className={`col-sm-2 control-label`}>
                name
              </label>
              <div className="col-sm-10">
                <input
                  ref={this.nameInputRef}
                  type="text"
                  className="form-control"
                  name="name1"
                  maxLength="45"
                  placeholder="At least 8 chars and no more than 45 chars"
                  size="31"
                  onChange={this.handleChange}
                  value={this.state.name}
                />
              </div>
              <div className="formError" ref={this.nameErrorRef}>
                Name shoud be at least 8 chars and no more than 45 chars
              </div>
            </div>
            <div className="form-group">
              <label htmlFor="email" className={`col-sm-2 control-label`}>
                email
              </label>
              <div className="col-sm-10">
                <input
                  ref={this.emailInputRef}
                  type="email"
                  className="form-control"
                  name="email"
                  maxLength="45"
                  placeholder="At least 8 chars and no more than 45 chars"
                  size="31"
                  onChange={this.handleChange}
                  value={this.state.email}
                />
              </div>
              <div className="formError" ref={this.emailErrorRef}>
                Email shoud be at least 8 chars and no more than 45 chars
              </div>
            </div>
            <div className="form-group">
              <label htmlFor="phone" className={`col-sm-2 control-label`}>
                phone
              </label>
              <div className="col-sm-10">
                <input
                  ref={this.phoneInputRef}
                  type="text"
                  className="form-control"
                  name="phone"
                  maxLength="45"
                  placeholder="At least 8 chars and no more than 30 chars"
                  size="31"
                  onChange={this.handleChange}
                  value={this.state.phone}
                />
              </div>
              <div className="formError" ref={this.phoneErrorRef}>
                Phone shoud be at least 8 chars and no more than 30 chars
              </div>
            </div>
            <div className="form-group">
              <label htmlFor="address" className={`col-sm-2 control-label`}>
                address
              </label>
              <div className="col-sm-10">
                <input
                  ref={this.addressInputRef}
                  type="text"
                  className="form-control"
                  name="address"
                  maxLength="45"
                  placeholder="At least 8 chars and no more than 45 chars"
                  size="31"
                  onChange={this.handleChange}
                  value={this.state.address}
                />
              </div>
              <div className="formError" ref={this.addressErrorRef}>
                Address shoud be at least 8 chars and no more than 45 chars
              </div>
            </div>
            <div className="form-group">
              <label htmlFor="creditcard" className={`col-sm-2 control-label`}>
                credit card
              </label>
              <div className="col-sm-10">
                <input
                  ref={this.ccInputRef}
                  type="text"
                  className="form-control"
                  name="creditcard"
                  maxLength="45"
                  placeholder="At least 16 chars and no more than 19 chars"
                  size="31"
                  onChange={this.handleChange}
                  value={this.state.creditcard}
                />
              </div>
              <div className="formError" ref={this.ccErrorRef}>
                Credit card number shoud be at least 16 chars and no more than
                19 chars
              </div>
            </div>
            <div className="form-group">
              <div className={`col-sm-offset-2 col-sm-10`}>
                <button type="submit" className={`btn btn-primary`}>
                  submit purchase
                </button>
              </div>
            </div>
          </form>
          <InfoBox>
            <ul>
              <li>Next-day delivery is guaranteed</li>
              <li>
                A &euro; {surcharge.toFixed(2)} delivery surcharge is applied to
                all purchase orders
              </li>
            </ul>
            <PriceBox>
              <tbody>
                <tr>
                  <PriceBoxTd>subtotal:</PriceBoxTd>
                  <PriceBoxSubTotalTd className="checkoutPriceColumn">
                    &euro; {cart.subtotal.toFixed(2)}
                  </PriceBoxSubTotalTd>
                </tr>
                <tr>
                  <PriceBoxTd>surcharge:</PriceBoxTd>
                  <PriceBoxSubTotalTd className="checkoutPriceColumn">
                    &euro; {surcharge.toFixed(2)}
                  </PriceBoxSubTotalTd>
                </tr>
                <tr>
                  <PriceBoxTd>total:</PriceBoxTd>
                  <PriceBoxSubTotalTd className="checkoutPriceColumn">
                    &euro; {(cart.subtotal + surcharge).toFixed(2)}
                  </PriceBoxSubTotalTd>
                </tr>
              </tbody>
            </PriceBox>
          </InfoBox>
        </div>
      </div>
    );
  }
}
