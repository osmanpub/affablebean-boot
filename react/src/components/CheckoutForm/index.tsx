import React from "react";
// import { types, useAlert } from "react-alert";
import { useForm } from "react-hook-form";
import { connect, useDispatch } from "react-redux";
import { Cart } from "../../interfaces/cart";
import { FormErrors } from "../../interfaces/ui";
import { purchaseOrder } from "../../net/checkout";
import { RootState } from "../../redux";
import { setFormErrors } from "../../redux/ui";
import "./CheckoutForm.css";
import {
  InfoBox,
  PriceBox,
  PriceBoxSubTotalTd,
  PriceBoxTd
} from "./CheckoutForm.styles";

type Props = {
  cart: Cart;
  formErrors: Array<FormErrors>;
  setFormErrors: Function;
};

type FormData = {
  address: string;
  creditCard: string;
  email: string;
  name: string;
  phone: string;
};

function CheckoutForm(props: Props) {
  const { cart, formErrors, setFormErrors } = props;
  // Problem with jest - https://github.com/schiehll/react-alert/issues/148
  // const alert = useAlert();
  const dispatch = useDispatch();
  const { register, handleSubmit, errors } = useForm<FormData>();

  const onSubmit = handleSubmit(
    ({ address, creditCard, email, name, phone }) => {
      dispatch(purchaseOrder({ address, creditCard, email, name, phone }));
    }
  );

  if (formErrors && Array.isArray(formErrors) && formErrors.length > 0) {
    let msg = "";

    formErrors.forEach(error => {
      msg += `Field "${error.param}" with value "${error.value}" has the following problem:\n"${error.msg}"`;
    });

    setFormErrors([]);

    alert(
      `There was a problem processing your order.\nPlease correct the following errors:\n${msg}`
    );

    // alert.show(
    //   `There was a problem processing your order.\nPlease correct the following errors:\n${msg}`,
    //   {
    //     timeout: 0,
    //     type: types.ERROR
    //   }
    // );
  }

  return (
    <div className="singleColumn">
      <h2>checkout</h2>
      <p data-cy="checkout-intro" data-testid="checkout-intro">
        In order to purchase the items in your shopping cart, please provide us
        with the following information:
      </p>
      <br />
      <form className="form-horizontal" onSubmit={onSubmit}>
        <div className="form-group">
          <label htmlFor="name" className={`col-sm-2 control-label`}>
            name
          </label>
          <div className="col-sm-10">
            <input
              className="form-control"
              data-cy="checkout-name"
              data-testid="checkout-name"
              name="name"
              placeholder="At least 3 chars and no more than 64 chars"
              size={32}
              ref={register({ required: true, minLength: 3, maxLength: 64 })}
              type="text"
            />
          </div>
          {errors.name && (
            <div className="formError">
              Name shoud be at least 3 chars and no more than 64 chars
            </div>
          )}
        </div>
        <div className="form-group">
          <label htmlFor="email" className={`col-sm-2 control-label`}>
            email
          </label>
          <div className="col-sm-10">
            <input
              className="form-control"
              data-cy="checkout-email"
              data-testid="checkout-email"
              name="email"
              placeholder="At least 8 chars and no more than 32 chars"
              size={32}
              ref={register({ required: true, minLength: 8, maxLength: 32 })}
              type="email"
            />
          </div>
          {errors.email && (
            <div className="formError">
              Email shoud be at least 8 chars and no more than 32 chars
            </div>
          )}
        </div>
        <div className="form-group">
          <label htmlFor="phone" className={`col-sm-2 control-label`}>
            phone
          </label>
          <div className="col-sm-10">
            <input
              className="form-control"
              data-cy="checkout-phone"
              data-testid="checkout-phone"
              name="phone"
              placeholder="At least 8 chars and no more than 32 chars"
              size={32}
              ref={register({ required: true, minLength: 8, maxLength: 32 })}
              type="text"
            />
          </div>
          {errors.email && (
            <div className="formError">
              Phone shoud be at least 8 chars and no more than 32 chars
            </div>
          )}
        </div>
        <div className="form-group">
          <label htmlFor="address" className={`col-sm-2 control-label`}>
            address
          </label>
          <div className="col-sm-10">
            <input
              className="form-control"
              data-cy="checkout-address"
              data-testid="checkout-address"
              name="address"
              placeholder="At least 8 chars and no more than 256 chars"
              size={32}
              ref={register({ required: true, minLength: 8, maxLength: 256 })}
              type="text"
            />
          </div>
          {errors.address && (
            <div className="formError">
              Address shoud be at least 8 chars and no more than 256 chars
            </div>
          )}
        </div>
        <div className="form-group">
          <label htmlFor="creditcard" className={`col-sm-2 control-label`}>
            credit card
          </label>
          <div className="col-sm-10">
            <input
              className="form-control"
              data-cy="checkout-cc"
              data-testid="checkout-cc"
              name="creditCard"
              placeholder="At least 16 chars and no more than 19 chars"
              size={32}
              ref={register({ required: true, minLength: 16, maxLength: 19 })}
              type="text"
            />
          </div>
          {errors.creditCard && (
            <div className="formError">
              Credit card number shoud be at least 16 chars and no more than 19
              chars
            </div>
          )}
        </div>
        <div className="form-group">
          <div className={`col-sm-offset-2 col-sm-10`}>
            <button
              className={`btn btn-primary`}
              data-cy="checkout-submit"
              type="submit"
            >
              submit purchase
            </button>
          </div>
        </div>
      </form>
      <InfoBox>
        <ul>
          <li>Next-day delivery is guaranteed</li>
          <li>
            A &euro; {surcharge.toFixed(2)} delivery surcharge is applied to all
            purchase orders
          </li>
        </ul>
        <PriceBox>
          <tbody>
            <tr>
              <PriceBoxTd>subtotal:</PriceBoxTd>
              <PriceBoxSubTotalTd className="checkoutPriceColumn">
                &euro; {cart.subtotal.toFixed(2)}
              </PriceBoxSubTotalTd>
            </tr>
            <tr>
              <PriceBoxTd>surcharge:</PriceBoxTd>
              <PriceBoxSubTotalTd className="checkoutPriceColumn">
                &euro; {surcharge.toFixed(2)}
              </PriceBoxSubTotalTd>
            </tr>
            <tr>
              <PriceBoxTd>total:</PriceBoxTd>
              <PriceBoxSubTotalTd className="checkoutPriceColumn">
                &euro; {(cart.subtotal + surcharge).toFixed(2)}
              </PriceBoxSubTotalTd>
            </tr>
          </tbody>
        </PriceBox>
      </InfoBox>
    </div>
  );
}

const surcharge = 3;

const mapStateToProps = (state: RootState) => ({
  cart: state.cart,
  formErrors: state.ui.formErrors
});

const mapDispatchToProps = {
  setFormErrors
};

export default connect(mapStateToProps, mapDispatchToProps)(CheckoutForm);
