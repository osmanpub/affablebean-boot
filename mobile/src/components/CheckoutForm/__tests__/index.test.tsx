import {act, fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import CheckoutForm from '..';

const mockStore = configureStore([]);

const initialState = {
  cart: {
    didInvalidate: false,
    isFetching: false,
    items: [],
    numberOfItems: 0,
    subtotal: 0,
  },
  purchase: {
    didInvalidate: false,
    isPosting: false,
    order: {},
  },
  ui: {
    formErrors: [],
  },
};

const store = mockStore(initialState);

const checkoutForm = (
  <Provider store={store}>
    <CheckoutForm />
  </Provider>
);

describe('<CheckoutForm />', () => {
  it('renders correctly', () => {
    const component = renderer.create(checkoutForm).toJSON();
    expect(component).toMatchSnapshot();
  });

  it('show intro', () => {
    const {getByTestId} = render(checkoutForm);
    const intro = getByTestId(/checkout-intro/i);
    expect(intro).toBeTruthy();
  });

  it('show 5 input fields', () => {
    const {getAllByTestId} = render(checkoutForm);
    const inputs = getAllByTestId('checkout-input');
    expect(inputs).toHaveLength(5);
  });

  it('show submit button', () => {
    const {getByTestId} = render(checkoutForm);
    const button = getByTestId(/checkout-submit/i);
    expect(button).toBeTruthy();
  });

  it("set name to 'osman'", () => {
    const {getByPlaceholderText} = render(checkoutForm);
    const input = getByPlaceholderText('Enter your name');

    act(() => {
      fireEvent.change(input, {nativeEvent: {text: 'osman'}});
    });

    // @ts-ignore
    expect(input.props.value).toBe('osman');
  });

  it("set phone to '1234567890'", () => {
    const {getByPlaceholderText} = render(checkoutForm);
    const input = getByPlaceholderText('Enter your phone number');

    act(() => {
      fireEvent.change(input, {nativeEvent: {text: '1234567890'}});
    });

    // @ts-ignore
    expect(input.props.value).toBe('1234567890');
  });

  it("set email to 'osman@gmail.com'", () => {
    const {getByPlaceholderText} = render(checkoutForm);
    const input = getByPlaceholderText('Enter your email address');

    act(() => {
      fireEvent.change(input, {nativeEvent: {text: 'osman@gmail.com'}});
    });

    // @ts-ignore
    expect(input.props.value).toBe('osman@gmail.com');
  });

  it("set address to 'nobody cares where you live'", () => {
    const {getByPlaceholderText} = render(checkoutForm);
    const input = getByPlaceholderText('Enter your address');

    act(() => {
      fireEvent.change(input, {
        nativeEvent: {text: 'nobody cares where you live'},
      });
    });

    // @ts-ignore
    expect(input.props.value).toBe('nobody cares where you live');
  });

  it("set cc to '1234567890123456'", () => {
    const {getByPlaceholderText} = render(checkoutForm);
    const input = getByPlaceholderText('Enter your credit card number');

    act(() => {
      fireEvent.change(input, {nativeEvent: {text: '1234567890123456'}});
    });

    // @ts-ignore
    expect(input.props.value).toBe('1234567890123456');
  });
});
