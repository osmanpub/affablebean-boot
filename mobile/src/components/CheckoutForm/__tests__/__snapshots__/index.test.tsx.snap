// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckoutForm /> renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "backgroundColor": "#f7f7e9",
        "padding": 16,
      }
    }
  >
    <Text>
      In order to purchase the items in your shopping cart, please provide us with the following information:
    </Text>
    <TextInput
      allowFontScaling={true}
      name="name"
      onChange={[Function]}
      placeholder="Enter your name"
      rejectResponderTermination={true}
      style={
        Object {
          "borderBottomColor": "black",
          "borderBottomWidth": 1,
          "padding": 8,
        }
      }
      underlineColorAndroid="transparent"
    />
    <TextInput
      allowFontScaling={true}
      name="email"
      onChange={[Function]}
      placeholder="Enter your email address"
      rejectResponderTermination={true}
      style={
        Object {
          "borderBottomColor": "black",
          "borderBottomWidth": 1,
          "padding": 8,
        }
      }
      underlineColorAndroid="transparent"
    />
    <TextInput
      allowFontScaling={true}
      name="phone"
      onChange={[Function]}
      placeholder="Enter your phone number"
      rejectResponderTermination={true}
      style={
        Object {
          "borderBottomColor": "black",
          "borderBottomWidth": 1,
          "padding": 8,
        }
      }
      underlineColorAndroid="transparent"
    />
    <TextInput
      allowFontScaling={true}
      name="address"
      onChange={[Function]}
      placeholder="Enter your address"
      rejectResponderTermination={true}
      style={
        Object {
          "borderBottomColor": "black",
          "borderBottomWidth": 1,
          "padding": 8,
        }
      }
      underlineColorAndroid="transparent"
    />
    <TextInput
      allowFontScaling={true}
      name="creditCard"
      onChange={[Function]}
      placeholder="Enter your credit card number"
      rejectResponderTermination={true}
      style={
        Object {
          "borderBottomColor": "black",
          "borderBottomWidth": 1,
          "padding": 8,
        }
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "paddingTop": 16,
        }
      }
    >
      Subtotal: € 
      0.00
    </Text>
    <Text>
      Surcharge: € 
      3.00
    </Text>
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Total: € 
      3.00
       
    </Text>
    <Text
      style={
        Object {
          "marginTop": 8,
        }
      }
    />
    <View
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Purchase
        </Text>
      </View>
    </View>
  </View>
</View>
`;
